lib/teleterm/gateway/gateway.go

```diff
@@ -32,19 +32,46 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
-// New creates an instance of Gateway
+// New creates an instance of Gateway. It starts a listener on the specified port but it doesn't
+// start the proxy – that's the job of Serve.
 func New(cfg Config) (*Gateway, error) {
 	if err := cfg.CheckAndSetDefaults(); err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	listener, err := net.Listen("tcp", fmt.Sprintf("%s:%s", cfg.LocalAddress, cfg.LocalPort))
+	result, err := newListenerAndLocalProxy(cfg, cfg.LocalPort)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```

Comment:
Most of the functionality of `New` and `SetLocalPort` is almost identical – they both need to create a listener on the given port and then initialize a local proxy with that listener.

The difference is that `New` then just returns a new instance of the `Gateway` struct while `SetLocalPort` mutates an existing instance, as well as stops the previous proxy and starts a new one.

I extracted the shared logic into `newListenerAndLocalProxy`.

Reply:
Hmm, good question. After thinking about it for a while I realized this: `RestartGateway` is a workaround which we're going to remove as soon as we implement gravitational/webapps.e#299. `gateway.Gateway` itself doesn't know how to refresh certs so it had to be done on the daemon level.

We could make `daemon.Service.SetGatewayLocalPort` create a new gateway under a different port and remove the old one but we'd need to pick either one of these three options:

1. Expose almost all fields on `gateway.Gateway` so that `daemon.Service` can copy them over. It'd need to do the equivalent of this but take all the values from the old gateway:

https://github.com/gravitational/teleport/blob/02b4f8575f5f0bc305a37c89b58e4469c2662921/lib/teleterm/clusters/cluster_gateways.go#L51-L63

2. Make `SetGatewayLocalPort` pull all the initialization data that `clusters.Cluster.CreateGateway` pulls (as shown above) without fetching new certs but then we'd still be making unnecessary calls to the cluster, for example to get the db so that we can then call `db.GetName()`.

https://github.com/gravitational/teleport/blob/02b4f8575f5f0bc305a37c89b58e4469c2662921/lib/teleterm/clusters/cluster_gateways.go#L42

3. Do a mix of 1 and 2, exposing some methods but not all just so that we can avoid those unnecessary calls.

In the end, currently `gateway.SetLocalPort` does only the absolutely necessary work to carry out the port change.

The combination of `RestartGateway` being a workaround and letting the struct manipulate its own internals makes me lean towards keeping `SetLocalPort` as it is.

What do you think?

Reply:
It's just I saw the `RestartGateway` function and wondered about that approach.

> The combination of RestartGateway being a workaround and letting the struct manipulate its own internals makes me lean towards keeping SetLocalPort as it is.

Yes, I think abstracting things inside the gateway like `SetLocalPort` is better with your explanations. We just need to make sure the local proxy swap inside the gateway is done properly. Let me do a detailed review 😄

Reply:
It's also kinda my fault for not documenting it better that `RestartGateway` is only a workaround – I mentioned it only in `lib/teleterm/api/proto/v1/service.proto` but people are less likely to look there than at the actual implementation.